### PR TITLE
Fix publish form not going away when done

### PR DESCRIPTION
**Pull request by shana**
Looks like the fix to #48 wasn't quite complete, @amytruong stumbled upon a regression of it in #157.

I tracked it down to the subscriptions going away when there's an error, because `NotifyDone` would be called as soon as the PublishRepository command was done, which would trigger the release of subscriptions. Unfortunately, if the publish error'd out, this would still trigger `NotifyDone`, so any further attempts at publishing would not have the listeners required to properly update the UI afterwards.

The reason we didn't notice this before is because the messages still show up - because those are being sent directly from the view model in GitHub.App into the VS Team Explorer notification area. This is a problem, they should be getting surfaced to the view and then displayed there. This PR doesn't change that, but it adds a bit of progress state reporting to start with (there's another PR around which surfaces error messages, I need to revive it)

Anyways, now it works as it should.

### CODE REVIEW

File: src/GitHub.App/Controllers/UIController.cs [hunk 1]
```diff
@@ -142,6 +136,25 @@ public IObservable<UserControl> SelectFlow(UIControllerFlow choice)
...
+        {
+            completion = new Subject<bool>();
+            return completion;
+        }
+
+        void End(bool success)
+        {
+            uiProvider.RemoveService(typeof(IConnection));
+            transition.OnCompleted();
+            completion.OnNext(success);
```
Comment (Haacked) [inline src/GitHub.App/Controllers/UIController.cs]: `completion` could be null `here` if you didn't call `ListenToCompletionState` first. Should it be an `InvalidOperationException` in that case?

Comment (shana) [inline src/GitHub.App/Controllers/UIController.cs]: It's an optional listener, so being null is fine, I'll just add the ? there that I forgot.

Comment (Haacked) [inline src/GitHub.App/Controllers/UIController.cs]: :thumbsup:

**Comment by shana** [comment]
Fixed null things, should be good now!

File: src/GitHub.App/Controllers/UIController.cs [hunk 2]
```diff
@@ -142,6 +136,27 @@ public IObservable<UserControl> SelectFlow(UIControllerFlow choice)
...
         }
 
+        /// <summary>
+        /// Allows listening to the completion state of the ui flow - whether
+        /// it was completed because it was cancelled or whether it succeeded.
+        /// </summary>
+        /// <returns>true for success, false for cancel</returns>
+        public IObservable<bool> ListenToCompletionState()
+        {
+            if (completion == null)
```
Comment (Haacked) [inline src/GitHub.App/Controllers/UIController.cs]: As a stylistic thing, I tend to like:

```csharp
completion = completion ?? new Subject<bool>();
```

But it's a personal preference. I'll still merge this.

**Pull request merged**